File: 2383-minimum-hours-of-training-to-win-a-competition/2383-minimum-hours-of-training-to-win-a-competition.ts
/**get total number of energy needed to win
    check if initial energy is not enough, if it isn't,
    add difference to answer, as one hour equals one unit energy
    given that we need to check experience level by level,
    check that our experience level is greater than current opponent,
    if no, get diff and train for (diff) hours(i.e add diff to answer)
    return ans
*/
function minNumberOfHours(initialEnergy: number, initialExperience: number, energy: number[], experience: number[]): number {
    const energy_sum = energy.reduce((acc, cur) => acc + cur, 0) + 1;
    let ans = energy_sum - initialEnergy > 0 ? energy_sum - initialEnergy : 0;
    
    let exp = initialExperience;
    for (let i = 0; i < energy.length; i++) {
        if(experience[i] >= exp){
            ans += experience[i] - exp + 1;
            exp += experience[i] - exp + 1;
        }
        exp += experience[i];
    }
    return ans
};

File: 2383-minimum-hours-of-training-to-win-a-competition/2383-minimum-hours-of-training-to-win-a-competition.py
class Solution:
    def minNumberOfHours(self, initialEnergy: int, initialExperience: int, energy: List[int], experience: List[int]) -> int:
        energy_sum = 1
        for e in energy:
            energy_sum += e
        ans = energy_sum - initialEnergy if energy_sum - initialEnergy > 0 else 0;
        
        exp = initialExperience;
        for i in range(len(energy)):
            if experience[i] >= exp:
                ans += experience[i] - exp + 1;
                exp += experience[i] - exp + 1;
            
            exp += experience[i];
        return ans
        